File: test/QueryBuilder.FSharp.Test/test_query_builder.fs
namespace QueryBuilder.FSharp.Test
open FsharpTest
open System
open Xunit

type test_query_builder() =
    let closeBracket = "%5d";
    let date = DateTime.UtcNow.ToString("yyyy-MM-dd");
    let dot = "%2c";
    let maxPrice = 20;
    let minPrice = 10;
    let openBracket = "%5b";
    let tickets = "tickets";
    
    let arrayEncode(array: int[]) =
        let arrayBody = 
            array 
            |> Seq.map(sprintf "%A")
            |> String.concat(dot)
        sprintf "%s%s%s" openBracket arrayBody closeBracket

    let createBaseQuery = 
        new FsharpTest.QueryBuilder()
        |> fun x -> x.WhereStartDateIs date
        |> fun x -> x.WhereEndDateIs date
        |> fun x -> x.WhereStartTicketPriceIs minPrice
        |> fun x -> x.WhereEndTicketPriceIs maxPrice
        |> fun x -> x.WhereTypeOfTicketsIs tickets

    [<Fact>]
    let ``when generating query without params should generate empty query``() =
        let finalQuery = 
            new FsharpTest.QueryBuilder()
            |> fun x -> x.CreateQuery()
        Assert.Equal(finalQuery, "", true)

    [<Fact>]
    let ``when generating query with null in arrays it should create query without this params``() =
        let organizers = [||];
        let categories = [||];
        let places = [||];
         
        let finalQuery =
            createBaseQuery
            |> fun x -> x.WhereOrganizerIs organizers
            |> fun x -> x.WhereCategoryIs categories
            |> fun x -> x.WherePlaceIs places
            |> fun x -> x.CreateQuery()

        let expectedResult = sprintf "start_date=%s&end_date=%s&startTicket=%d&endTicket=%d&tickets=%s" date date minPrice maxPrice tickets
        Assert.Equal(finalQuery, expectedResult, true)

    [<Fact>]
    let ``when generating query it should create valid query``() =
        let organizers = [| 2; 3 |];
        let categories = [| 5; 6 |];
        let places = [| 7; 4 |];
         
        let finalQuery =
            createBaseQuery
            |> fun x -> x.WhereOrganizerIs organizers
            |> fun x -> x.WhereCategoryIs categories
            |> fun x -> x.WherePlaceIs places
            |> fun x -> x.CreateQuery()

        let expectedResult = 
            sprintf "start_date=%s&end_date=%s&startTicket=%d&endTicket=%d&tickets=%s&organizer=%s&category=%s&place=%s" 
                date date minPrice maxPrice tickets (arrayEncode organizers) (arrayEncode categories) (arrayEncode places)
        Assert.Equal(finalQuery, expectedResult, true)

File: test/QueryBuilder.Test/test_query_builder.cs
using Shouldly;
using System;
using Xunit;

namespace QueryBuilder.Test
{
    public class test_query_builder
    {
        private const int minPrice = 10;
        private const int maxPrice = 20;
        private const string tickets = "tickets";
        private readonly string date;
        private const string dot = "%2c";
        private const string openBracket = "%5b";
        private const string closeBracket = "%5d";

        public test_query_builder()
        {
            date = DateTime.UtcNow.ToString("yyyy-MM-dd");
        }

        [Fact]
        public void should_create_emtpy_query()
        {
            var query = new FsharpTest.QueryBuilder();

            var finalQuery = query.CreateQuery();

            finalQuery.ShouldBe("");
        }

        [Fact]
        public void should_return_valid_criteria()
        {
            var organizers = new [] { 2, 3 };
            var categories = new [] { 5, 6 };
            var places = new [] { 7, 4 };

            var query = CreateBaseQuery()
                .WhereOrganizerIs(organizers)
                .WhereCategoryIs(categories)
                .WherePlaceIs(places);

            var finalQuery = query.CreateQuery();

            finalQuery.ShouldBe($"start_date={date}&" +
                                $"end_date={date}&" +
                                $"startTicket={minPrice}&" +
                                $"endTicket={maxPrice}&" +
                                $"tickets={tickets}&" +
                                $"organizer={openBracket}{string.Join(dot, organizers)}{closeBracket}&" +
                                $"category={openBracket}{string.Join(dot, categories)}{closeBracket}&" +
                                $"place={openBracket}{string.Join(dot, places)}{closeBracket}");
        }

        private FsharpTest.QueryBuilder CreateBaseQuery()
        {
            return new FsharpTest.QueryBuilder()
                .WhereStartDateIs(date)
                .WhereEndDateIs(date)
                .WhereStartTicketPriceIs(minPrice)
                .WhereEndTicketPriceIs(maxPrice)
                .WhereTypeOfTicketsIs(tickets);
        }

        [Fact]
        public void should_return_valid_criteria_when_passing_null_values()
        {
            var organizers = new int[] {  };
            var categories = new int[] {  };
            var places = new int[] {  };

            var query = CreateBaseQuery()
                .WhereOrganizerIs(organizers)
                .WhereCategoryIs(categories)
                .WherePlaceIs(places);

            var finalQuery = query.CreateQuery();

            finalQuery.ShouldBe($"start_date={date}&" +
                                   $"end_date={date}&" +
                                   $"startTicket={minPrice}&" +
                                   $"endTicket={maxPrice}&" +
                                   $"tickets={tickets}");
        }
    }
}
